stdlib/src/tempfile/tempfile.mojo

```diff
@@ -239,3 +239,137 @@ struct TemporaryDirectory:
             return True
         except:
             return False
+
+
+struct NamedTemporaryFile:
+    """A handle to a temporary file."""
+
+    var _file_handle: FileHandle
+    """The underlying file handle."""
+    var _delete: Bool
+    """Whether the file is deleted on close."""
+    var name: String
+    """Name of the file."""
+
+    fn __init__(
+        inout self,
+        mode: String = "w",
+        suffix: String = "",
+        prefix: String = "tmp",
+        dir: Optional[String] = None,
+        delete: Bool = True,
+    ) raises:
+        """Create a named temporary file. Can be used as a context manager.
+        This is a wrapper around a `FileHandle`,
+        os.remove is called in close method if `delete` is True.
+
+        Args:
+            mode: The mode to open the file in (the mode can be "r" or "w").
+            suffix: Suffix to use for the file name.
+            prefix: Prefix to use for the file name.
+            dir: Directory in which the file will be created.
+            delete: Whether the file is deleted on close.
+        """
+
+        var final_dir = Path(dir.value()) if dir else Path(
+            _get_default_tempdir()
+        )
+
+        self._delete = delete
+        self.name = ""
+
+        for _ in range(TMP_MAX):
+            var potential_name = final_dir / (
+                prefix + _get_random_name() + suffix
+            )
+            if os.path.exists(potential_name):
+                continue
+            try:
+                # TODO for now this name could be relative,
+                # python implementation expands the path,
+                # but several functions are not yet implemented in mojo
+                # i.e. abspath, normpath
+                self.name = potential_name.__fspath__()
+                self._file_handle = FileHandle(
+                    potential_name.__fspath__(), mode=mode
+                )
+                return
+            except:
+                continue
+        else:
+            raise Error("Failed to create temporary file")
+
+    @always_inline
+    fn __del__(owned self):
+        """Closes the file handle."""
+        try:
+            self.close()
+        except:
+            pass
+
+    fn close(inout self) raises:
+        """Closes the file handle."""
+        self._file_handle.close()
+        if self._delete:
+            os.remove(self.name)
+
+    fn __moveinit__(inout self, owned existing: Self):
+        """Moves constructor for the file handle.
+
+        Args:
+          existing: The existing file handle.
+        """
+        self._file_handle = existing._file_handle^
+        self._delete = existing._delete
+        self.name = existing.name^
+
+    @always_inline
+    fn read(self, size: Int64 = -1) raises -> String:
```

Comment:
Q: Just to confirm, providing these wrappers is necessary because we need to allow this to be used both within a context manager and as a normal object, right? (And also because in python this is actually a function).

This is fine, but I wonder if we could introduce a trait for things that implement these FileHandle-like interfaces.

Reply:
yes, I don't know if there is a better way to do this in mojo for now, just as a reference the python implementation does not wraps all the methods of the underlying object, it reimplements  `__getattr__` to directly return the methods of the real file object: https://github.com/python/cpython/blob/e83ce850f433fd8bbf8ff4e8d7649b942639db31/Lib/tempfile.py#L489-L506

I guess it would make sense to define some traits for IO stuff, the equivalent of the clases defined here: https://docs.python.org/3/library/io.html